File: clients/windows/Mailcal/Services/OutboxRows.cs
// The Outbox list's one decision: what each queued send says about itself (docs/sending.md).
//
// WinUI-free and L10n-free on purpose, so Mailcal.Tests can link it. The failure it exists to
// catch is silent in the running app: a state mapped to the wrong word tells someone their
// message is waiting when it is already on its way, or offers "Send now" on a message that may
// have been delivered, which is how it arrives twice. Neither looks wrong on screen.

using System.Collections.Generic;
using Allodia.Mailcal.ViewModels;
using uniffi.mailcal_bindings;

namespace Allodia.Mailcal.Services;

/// <summary>What each queued state says and draws, supplied by the shell.</summary>
/// <param name="Waiting">"Waiting to send": the ordinary state, stated plainly.</param>
/// <param name="Sending">"Sending…": a submission is in flight.</param>
/// <param name="Unconfirmed">"Delivery not confirmed": it may already have arrived.</param>
/// <param name="WaitingGlyph">The Segoe Fluent glyph for <paramref name="Waiting"/>.</param>
/// <param name="SendingGlyph">The glyph for <paramref name="Sending"/>.</param>
/// <param name="UnconfirmedGlyph">The glyph for <paramref name="Unconfirmed"/>. The only warning
/// of the three: it is the one state a person may need to check on another device.</param>
/// <param name="NoSubject">What stands in for a subject the message does not have. The same words
/// the message list and the reading header use, because it is the same absence.</param>
/// <remarks>
/// The words arrive as parameters because <c>L10n</c> needs a Windows TFM that Mailcal.Tests does
/// not have, and the glyphs because the Segoe Fluent codepoints live in the shell
/// (<c>MainWindow.Sidebar.cs</c> says why they are written as escapes).
/// </remarks>
public readonly record struct OutboxLabels(
    string Waiting,
    string Sending,
    string Unconfirmed,
    string WaitingGlyph,
    string SendingGlyph,
    string UnconfirmedGlyph,
    string NoSubject);

/// <summary>Projects the core's queued sends into the rows the Outbox list draws.</summary>
/// <remarks>
/// Internal, not public: it reads the generated <c>QueuedRow</c>, which UniFFI emits as an
/// internal type. <see cref="QueuedRowItem"/> stays public, because the XAML binds it.
/// </remarks>
internal static class OutboxRows
{
    /// <summary>
    /// The Outbox's rows, in the order the core gave them: oldest first, which is the order they
    /// will go out in.
    /// </summary>
    /// <param name="queued">The snapshot's queued sends.</param>
    /// <param name="labels">The state words and glyphs.</param>
    /// <param name="accountEmail">An account id resolved to the address the user knows it by. Every
    /// row draws it (docs/folder-pane.md, rule 18), because this list holds every account's mail at
    /// once and nothing else on the row says which one a message is waiting on.</param>
    public static List<QueuedRowItem> Build(
        IReadOnlyList<QueuedRow> queued,
        OutboxLabels labels,
        Func<string, string> accountEmail)
    {
        var rows = new List<QueuedRowItem>(queued.Count);
        foreach (var row in queued)
        {
            var state = StateOf(row.State);
            var account = accountEmail(row.Account);
            rows.Add(new QueuedRowItem
            {
                Account = row.Account,
                Op = row.Op,
                // Both lines can genuinely come back empty, and an empty line reads as a rendering
                // fault rather than as a fact. A blank subject is ordinary mail, which is why there
                // are already words for it; an empty To is a message addressed by Bcc alone, and
                // the account it goes from is then the only thing known about where it is going.
                ToText = row.To.Length > 0 ? row.To : account,
                SubjectText = row.Subject.Length > 0 ? row.Subject : labels.NoSubject,
                AccountText = account,
                State = state,
                StateText = TextFor(state, labels),
                StateGlyph = GlyphFor(state, labels),
            });
        }
        return rows;
    }

    /// <summary>The mirror of the core's state onto the one the view models carry.</summary>
    public static QueuedSendState StateOf(QueuedState state) => state switch
    {
        QueuedState.Sending => QueuedSendState.Sending,
        QueuedState.Unconfirmed => QueuedSendState.Unconfirmed,
        // Waiting, and anything a later core adds: the plain, non-alarming reading, and the only
        // one of the three that offers the actions. A new state arriving here as actionable would
        // be worse than arriving as waiting, so the discard falls this way deliberately.
        _ => QueuedSendState.Waiting,
    };

    private static string TextFor(QueuedSendState state, OutboxLabels labels) => state switch
    {
        QueuedSendState.Sending => labels.Sending,
        QueuedSendState.Unconfirmed => labels.Unconfirmed,
        _ => labels.Waiting,
    };

    private static string GlyphFor(QueuedSendState state, OutboxLabels labels) => state switch
    {
        QueuedSendState.Sending => labels.SendingGlyph,
        QueuedSendState.Unconfirmed => labels.UnconfirmedGlyph,
        _ => labels.WaitingGlyph,
    };
}

File: clients/android/app/src/main/java/eu/allodia/mailcal/OutboxRows.kt
// The Outbox list's one decision: what each queued send says about itself, and whether it may be
// acted on at all (docs/sending.md).
//
// Compose-free on purpose, so the JVM suite can pin it without composing a screen. The failure it
// exists to catch is silent on the screen: a state mapped to the wrong word tells someone their
// message is waiting when it is already on its way, and offering "Send now" on a message that may
// have been delivered is how it arrives twice. Neither looks wrong.
package eu.allodia.mailcal

import android.content.Context
import uniffi.mailcal_bindings.AccountRow
import uniffi.mailcal_bindings.QueuedRow
import uniffi.mailcal_bindings.QueuedState

/** One unsent message, as the Outbox list draws it. */
internal data class OutboxRowItem(
    /** The account it will be sent from, and the op id an action names together with it. */
    val account: String,
    val op: ULong,
    /** The recipients, comma-joined, or the account when the message is addressed by Bcc alone. */
    val recipients: String,
    /** The subject, or the words the message list uses for mail that has none. */
    val subject: String,
    /** Where the send has got to, in the app's own words. */
    val stateText: String,
    /** The address of the account it goes out from, drawn on every row. */
    val accountText: String,
    /**
     * Whether this row offers its three actions at all.
     *
     * A message in flight is on its way and cannot be called back; one awaiting confirmation may
     * already be in front of its recipients, and offering to send that again is how it arrives
     * twice (docs/sending.md). Both show their state and offer nothing.
     */
    val actionable: Boolean,
)

/**
 * Projects the core's queued sends into the rows the Outbox draws, in the order the core gave
 * them: oldest first, which is the order they will go out in.
 *
 * `accounts` resolves an account id to the address the user knows it by. Every row draws it
 * (docs/folder-pane.md, rule 18), because this list is the one place in the app holding every
 * account's mail at once and nothing else on the row says which identity a message is waiting on.
 */
internal fun outboxRows(
    queued: List<QueuedRow>,
    accounts: List<AccountRow>,
    ctx: Context,
): List<OutboxRowItem> {
    val emails = accounts.associate { it.id to it.email }
    return queued.map { row ->
        // An account the user has since removed still names something rather than nothing.
        val account = emails[row.account] ?: row.account
        OutboxRowItem(
            account = row.account,
            op = row.op,
            // Both lines can genuinely come back empty, and an empty line reads as a rendering
            // fault rather than as a fact. A blank subject is ordinary mail, which is why there
            // are already words for it; an empty To is a message addressed by Bcc alone, and the
            // account it goes from is then the only thing known about where it is going.
            recipients = row.to.ifEmpty { account },
            subject = row.subject.ifEmpty { L10n.mail_no_subject(ctx) },
            stateText = queuedStateText(row.state, ctx),
            accountText = account,
            actionable = isActionable(row.state),
        )
    }
}

/** What a queued send's state is called on screen. */
internal fun queuedStateText(state: QueuedState, ctx: Context): String = when (state) {
    QueuedState.SENDING -> L10n.outbox_sending(ctx)
    QueuedState.UNCONFIRMED -> L10n.outbox_unconfirmed(ctx)
    QueuedState.WAITING -> L10n.outbox_waiting(ctx)
}

/**
 * Whether a queued send may be acted on.
 *
 * Only `WAITING`. A state a later core adds arrives here as not actionable, which is the safe
 * way for this to fall: a new state read as actionable would offer a retry on a message nobody
 * has decided is safe to retry.
 */
internal fun isActionable(state: QueuedState): Boolean = state == QueuedState.WAITING
